bindings/profilers/wall.cc

```diff
@@ -25,213 +27,460 @@
-           Nan::New<Number>(profile->GetStartTime()));
-  Nan::Set(js_profile,
-           Nan::New<String>("endTime").ToLocalChecked(),
-           Nan::New<Number>(profile->GetEndTime()));
-  return js_profile;
+      sample->GetChildrenCount() == 0 &&
+      strncmp("(idle)", sample->GetFunctionNameStr(), 7) == 0;
+}
+
+LabelSetsByNode WallProfiler::GetLabelSetsByNode(CpuProfile* profile,
+                                                 uint64_t startTime) {
+  LabelSetsByNode labelSetsByNode;
+
+  if (contexts.empty() || profile->GetSamplesCount() == 0) {
+    return labelSetsByNode;
+  }
+  auto isolate = Isolate::GetCurrent();
+  SampleContext sampleContext = contexts.pop_front();
+
+  auto sampleCount = profile->GetSamplesCount();
+  uint64_t halfInterval =
+      (sampleCount > 1 ? (profile->GetSampleTimestamp(sampleCount - 1) -
+                          profile->GetSampleTimestamp(0)) /
+                             (sampleCount - 1)
+                       : samplingInterval) /
+      2;
+
+  // Assumption is that startTime and profile->GetSampleTimestamp(0) were both
+  // taken very closely to each other in real time so they can be used as a
+  // common point of time origin, since startTime is taken in
+  // WallProfiler::Start just after calling WallProfiler::StartImpl.
+  // Sample times are in micros, context times are
+  // in nanos. We are using these values - adjusted for half interval in the
+  // past so we don't suffer any overflows anywhere in the arithmetic below.
+  int64_t zeroSampleTime = profile->GetSampleTimestamp(0) - halfInterval;
+  uint64_t zeroContextTime = startTime - halfInterval * 1000;
+
+  for (int i = 0; i < sampleCount; i++) {
+    auto sample = profile->GetSample(i);
+    if (isIdleSample(sample)) {
+      continue;
+    }
+    int64_t sampleTimestamp = profile->GetSampleTimestamp(i) - zeroSampleTime;
+    // Compute earliest (inclusive) and latest (exclusive) context timestamps
+    // that can belong to this sample. Use half the distance to neighboring
+    // samples, or barring that half the sampling interval
+    uint64_t earliest =
+        1000 * (i > 0 ? (sampleTimestamp + (profile->GetSampleTimestamp(i - 1) -
+                                            zeroSampleTime)) /
+                            2
+                      : (sampleTimestamp - halfInterval));
+    uint64_t latest =
+        1000 * (i < sampleCount - 1
+                    ? (sampleTimestamp +
+                       (profile->GetSampleTimestamp(i + 1) - zeroSampleTime)) /
+                          2
+                    : (sampleTimestamp + halfInterval));
+
+    // This loop will drop all contexts that are too old to be associated with
+    // the current sample; associate those (normally one) that are close enough
+    // to it in time, and stop as soon as it sees a context that's too recent
+    // for this sample.
+    for (;;) {
+      auto contextTimestamp = sampleContext.timestamp - zeroContextTime;
```

Comment:
`contextTimestamp` is an `uint64_t` (because `sampleContext.timestamp` and `zeroContextTime` are both `uint64_t), but `zeroContextTime` might be bigger than `sampleContext.timestamp` (if a context from a previous period was not matched for example), leading to a `contextTimestamp` holding a very large value and test `(contextTimestamp < earliest)` failing whereas is should.
I find it best to avoid unsigned arithmetic types in c++, because they lead to surprising behaviours in the face of integer promotion rules (eg. `-1 > 1u` being true).

Reply:
Our only source of unsigned ints was `uv_hrtime()` but fortunately we could now completely ditch it in favor of `v8::base::TimeTicks` so this'll no longer be an issue 🎉